include/pthread.h

```diff
@@ -0,0 +1,376 @@
+	s32_t tv_nsec;
+};
+
+static inline s32_t _ts_to_ms(const struct timespec *to)
+{
+	return (to->tv_sec * 1000) + (to->tv_nsec / 1000000);
+}
+
+typedef struct pthread_mutex {
+	struct k_sem *sem;
+} pthread_mutex_t;
+
+typedef struct pthread_mutexattr {
+	int unused;
+} pthread_mutexattr_t;
+
+typedef struct pthread_cond {
+	_wait_q_t wait_q;
+} pthread_cond_t;
+
+typedef struct pthread_condattr {
+	int unused;
+} pthread_condattr_t;
+
+/**
+ * @brief Declare a pthread condition variable
+ *
+ * Declaration API for a pthread condition variable.  This is not a
+ * POSIX API, it's provided to better conform with Zephyr's allocation
+ * strategies for kernel objects.
+ *
+ * @param name Symbol name of the condition variable
+ */
+#define PTHREAD_COND_DEFINE(name)					\
+	struct pthread_cond name = {					\
+		.wait_q = SYS_DLIST_STATIC_INIT(&name.wait_q),		\
+	}
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+static inline int pthread_cond_init(pthread_cond_t *cv,
+				    const pthread_condattr_t *att)
+{
+	ARG_UNUSED(att);
+	sys_dlist_init(&cv->wait_q);
+	return 0;
+}
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+static inline int pthread_cond_destroy(pthread_cond_t *cv)
+{
+	return 0;
+}
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+int pthread_cond_signal(pthread_cond_t *cv);
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+int pthread_cond_broadcast(pthread_cond_t *cv);
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+int pthread_cond_wait(pthread_cond_t *cv, pthread_mutex_t *mut);
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+int pthread_cond_timedwait(pthread_cond_t *cv, pthread_mutex_t *mut,
+			   const struct timespec *to);
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1.
+ *
+ * Note that pthread attribute structs are currently noops in Zephyr.
+ */
+static inline int pthread_condattr_init(pthread_condattr_t *att)
+{
+	return 0;
+}
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ *
+ * Note that pthread attribute structs are currently noops in Zephyr.
+ */
+static inline int pthread_condattr_destroy(pthread_condattr_t *att)
+{
+	return 0;
+}
+
+/**
+ * @brief Declare a pthread mutex
+ *
+ * Declaration API for a pthread mutex.  This is not a POSIX API, it's
+ * provided to better conform with Zephyr's allocation strategies for
+ * kernel objects.
+ *
+ * @param name Symbol name of the mutex
+ */
+#define PTHREAD_MUTEX_DEFINE(name)		\
+	K_SEM_DEFINE(name##_psem, 1, 1);	\
+	struct pthread_mutex name = {		\
+		.sem = &name##_psem,		\
+	}
+
+
+/**
+ * @brief POSIX threading compatibility API
+ *
+ * See IEEE 1003.1
+ */
+static inline int pthread_mutex_init(pthread_mutex_t *m,
+				     const pthread_mutexattr_t *att)
+{
+	ARG_UNUSED(att);
+
+	k_sem_init(m->sem, 1, 1);
```

Comment:
Not an issue with this patch, but I've seen plenty of places where `k_sem_init(sem, 0, 1); k_sem_take(sem);` has been used instead. Maybe a Coccinelle script should be written to initialize the semaphore taken in all those places in a single pass?